### PR TITLE
Bug fix: the `..` disappear.

### PR DESCRIPTION
**Pull request by pigletDips**
Avoid the root inode use 0 as nid, which may cause the `..` directory disappear in root dir 's subdirs; What's more, change the dirname's length from in character to in byte.

Signed-off-by: Qi Wang <mpiglet@outlook.com>

### CODE REVIEW

**Review by jiangliu — APPROVED** [review]

**Comment by jiangliu** [comment]
Fixes: #403

**Review by liubogithub — COMMENTED** [review]
Can you please make the commit log even more verbose given the problem is a bit tricky?

As it fixes a problem in nydus-image binary, a proper title can be `"nydus-image: fix xxx problem"`.

In the description part, please explain why the file of "inode id == 0" doesn't get shown by glibc's `getdents`/`readdir` wrapper (a link for reference would be appreciated).

File: src/bin/nydus-image/core/node.rs [hunk 1]
```diff
@@ -600,8 +600,8 @@ impl Node {
                         entry.set_name_offset(nameoff as u16);
                         dir_data.extend(entry.as_ref());
                         entry_names.push(*name);
-
-                        nameoff += name.len() as u64;
+                        // Use length in byte, instead of length in character.
```
Comment (liubogithub) [inline src/bin/nydus-image/core/node.rs:603]: Please make a separate patch for this part.  They look irrelevant to the title.

File: src/bin/nydus-image/core/node.rs [hunk 2]
```diff
@@ -1073,13 +1073,15 @@ impl Node {
...
+        // Use length in byte, instead of length in character.
+        let mut d_size: u64 = (".".as_bytes().len()
+            + size_of::<RafsV6Dirent>()
+            + "..".as_bytes().len()
+            + size_of::<RafsV6Dirent>()) as u64;
 
         // Safe as we have check tree above.
         for child in tree.children.iter() {
-            let len = child.node.name().len() + size_of::<RafsV6Dirent>();
+            let len = child.node.name().as_bytes().len() + size_of::<RafsV6Dirent>();
```
Comment (liubogithub) [inline src/bin/nydus-image/core/node.rs:1084]: Ditto.

File: src/bin/nydus-image/core/bootstrap.rs [hunk 3]
```diff
@@ -560,7 +560,8 @@ impl Bootstrap {
             (0, 0)
         };
 
-        let orig_meta_addr = bootstrap_ctx.nodes[0].offset;
+        // To avoid root inode use nid 0
```
Comment (liubogithub) [inline src/bin/nydus-image/core/bootstrap.rs:563]: Please explain the 'why to avoid' if possible.

**Review by liubogithub — CHANGES_REQUESTED** [review]

File: src/bin/nydus-image/core/bootstrap.rs [hunk 4]
```diff
@@ -560,7 +560,8 @@ impl Bootstrap {
...
+        let orig_meta_addr = bootstrap_ctx.nodes[0].offset - EROFS_INODE_SLOT_SIZE as u64;
```
[the remark this answers is not in the document]
Comment (pigletDips) [reply src/bin/nydus-image/core/bootstrap.rs:564]:
```rust
let orig_meta_addr = bootstrap_ctx.nodes[0].offset;
let root_nid = calculate_nid(
            bootstrap_ctx.nodes[0].offset - orig_meta_addr + meta_addr,
            meta_addr,
        );

pub fn calculate_nid(offset: u64, meta_size: u64) -> u64 {
    (offset - meta_size) >> EROFS_INODE_SLOT_BITS
}
```
If we just set meta_addr to meta_addr - EROFS_BLOCK_SIZE, I think root_nid will still be 0.

[the remark this answers is not in the document]
Comment (hsiangkao) [reply src/bin/nydus-image/core/bootstrap.rs:564]: Take an example for explaination...
If the root inode starts at the block 1, if we set on-disk meta_blkaddr = 1, so the nid would be 0;
But if we set on-disk meta_blkaddr = 0, so the nid would be = 4096/32 = 128 instead.

File: src/bin/nydus-image/core/bootstrap.rs [hunk 5]
```diff
@@ -560,7 +560,10 @@ impl Bootstrap {
             (0, 0)
         };
 
-        let orig_meta_addr = bootstrap_ctx.nodes[0].offset;
+        // Leave one block between the superblock and the inodes
+        // to avoid root inode use 0 as nid,
+        // which may cause the `..` disappear in some subdirs of root directory.
```
Comment (bergwolf) [inline src/bin/nydus-image/core/bootstrap.rs:565]: The "WHY" is not really clear to me. Why would it cause `..` to disappear in some subdirs of the root directory?

Also when the comments says `Leave one block between the superblock and the inodes`, the meta address is moved one block backward. Wouldn't it cause bootstrap to overlap with the inode table?

Comment (bergwolf) [inline src/bin/nydus-image/core/bootstrap.rs:565]: The "WHY" is not really clear to me. Why would it cause `..` to disappear in some subdirs of the root directory?

Also when the comments says `Leave one block between the superblock and the inodes`, the meta address is moved one block backward. Wouldn't it cause bootstrap to overlap with the inode table?

Comment (hsiangkao) [reply src/bin/nydus-image/core/bootstrap.rs:565]: yeah, actually I understand what it really means. but the commit might need do some refinement anyway.

**Review by liubogithub — COMMENTED** [review]
For all contributors's reference as well,  the format of a typical bugfix format is
```
$(subsystem): $(what is this patch for)

$(explain why the following changes are needed UNLESS they are way too simple to understand)

Signed-off-by: 
```

The purpose of doing above is to help people better follow what happened here.

As for this patch, please explain why `as_bytes()` is preferred or what corner case would be hit without the patch.

**Review by liubogithub — APPROVED** [review]

**Review by hsiangkao — APPROVED** [review]
